streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -135,31 +189,42 @@ public void maybeWaitForNonEmptyTopology(final Supplier<State> threadState) {
         }
     }
 
-    public void registerAndBuildNewTopology(final InternalTopologyBuilder newTopologyBuilder) {
+    /**
+     * Adds the topology and registers a future that listens for all threads on the older version to see the update
+     */
+    public KafkaFuture<Void> registerAndBuildNewTopology(final InternalTopologyBuilder newTopologyBuilder) {
+        final KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
         try {
             lock();
             version.topologyVersion.incrementAndGet();
             log.info("Adding NamedTopology {}, latest topology version is {}", newTopologyBuilder.topologyName(), version.topologyVersion.get());
+            version.activeTopologyWaiters.add(new TopologyVersionWaiters(topologyVersion(), future));
```

Comment:
Please see my other comment: instead of keeping a single list of `topologyVersionWaiters`, could we just keep a list of futures per thread, and once updated immediately complete all of them?

Reply:
responded to your other comment :)